### ISSUE TITLE
1.0 roadmap

### ISSUE THREAD

**Issue by plandem** [tommikaikkonen/redux-orm#140]
Let's discuss what we needed to release a v1

- [ ] refactor many-many field declaration to allow forward/backward declaration at each Model implicit (almost like now, but with ability to declare relation at target model too)/explicit (without any internal magic, more clean code, more easy to support)
- [ ] indexes for  any attributes to speedup queries

what else?
@tiii @markerikson @tommikaikkonen @NathanBWaters

**Comment by tiii** [comment]
Up to date documentation! 😉 

Keep relation ids in a seperate field instead of overloading it with the relation  [#102]
Computed fields via attr descriptor?

And since i guess everybody using redux-orm is doing implementing this in some way:
A simple API for denormalizing (including computed fields / relations etc.)?

**Comment by nealoke** [comment]
@tiii +1 for the documentation and api for denormalizing (including computed fields / relations). Great to see that this library is getting active again!

**Comment by plandem** [comment]
What do you mean denormalization API in our case?!

**Comment by nealoke** [comment]
@plandem I understood it as this.

Currently when you want to get something out of the `orm` in a `selector` with it's relations (in this case questions) you need to do the following
```
export const getAccounts = createSelector(orm, getOrm, (session) => {
	const accounts = session.Account.all().toModelArray();
	const enhancedAccounts = [];

	accounts.map(accountModel => {
		enhancedAccounts.push({
			...accountModel.ref,
			questions: accountModel.questions.toRefArray()
		})
	});

	return enhancedAccounts;
});
```


While in the ideal case you could do something like:
```
export const getAccounts = createSelector(orm, getOrm, (session) => {
	return session.Account.all().toRefArrayWithConnections(); // with a different name for the function
});
```

**Comment by plandem** [comment]
Well, we are using this library a lot in our application and I must say, that your case is very limited. E.g. You need only subset of relations where some just array of id, but other - array of reduced version of related model(e.g. Id, title), another one must be indexed by some index field. So many cases. In our application we just created few own createSelectors for most cases with some options to tune selectors.

Probably it's better to create another helper-library for it.

**Comment by nealoke** [comment]
@plandem yhea maybe that's better. I just think it is a nice convenience. I'm also using this in my application and I find myself doing this quite often though. 

Maybe @tiii has more use cases or another insight on this? :)

**Comment by plandem** [comment]
@nealoke in our case we created and using next selectors:
```js
createSelector, //unified selector, that creates ORM selector or selector from reselect
createRecordSelector, //to select single record with query info for relations and etc
createListSelector, //to select all records with query info for relations and etc
createTreeSelector, //to select all records resolved as tree with query info for relations and etc
createGroupedSelector, //to select all records grouped by field, with query info for relations and etc
createMapSelector,//to select all records indexed by id with query info for relations and etc
```

All these selectors has set of options to tune result. So as I said, your case is very limited. And probably external helper-library would be better. Maybe I will public our version.

**Comment by tiii** [comment]
@plandem @nealoke

I sometimes request multiple nested resources so the selectors to grab them get kind of nasty. But thats rather a problem of poor design on my side. I guess I should use more redux containers.

Nonetheless I am also interested in your selectors and I would say they already seem like some easier API for denormalizing. Why move them to a seperate library?

**Comment by plandem** [comment]
@tiii separate library, because I'm not sure that it fits here (E.g. in our case we use multi-schema ORM, where some schemas are same but has differ keys at global state). For me it's **selector's** area. Anyway, right now I'm not ready to make it public yet. Need to revise it :)

>I guess I should use more redux containers
I recommend you to look for this library: https://github.com/artsy/react-redux-controller

It's small and helped me to reduce complexity of application.

**Comment by nealoke** [comment]
@plandem Are you also planning to update the [redux-orm-propTypes](https://github.com/tommikaikkonen/redux-orm-proptypes)? Or do you think this package is obsolete?

**Comment by plandem** [comment]
I don't use it, so probably not. Maybe someone else

**Comment by NathanBWaters** [comment]
@plandem Only other thing I could think of would be performance testing and potentially incorporating Flow.  I don't believe either of them are required for v1 though.

**Comment by NathanBWaters** [comment]
@nealoke I'll take on updating https://github.com/tommikaikkonen/redux-orm-proptypes because I want to use it haha.

**Comment by plandem** [comment]
@NathanBWaters agree, that someday we need to tune performance :)

**Comment by Aryk** [comment]
hey @plandem  - just wanted to throw in my 2 cents...regarding " refactor many-many field declaration to allow forward/backward declaration at each Model"

If many to many has forward/backward declaration would the normal foreignkey/hasmany relationships also need to be defined separately on each model as well.

I think having the backwards relationship explicitly defined would be better. I currently have to keep comments in my code alerting other developers that "hey the backwards relationship is on the other model..." :)

**Comment by plandem** [comment]
@Aryk yep, if explicitly then you will have to define on both models and in case of custom **through** model it can be verbose :) but...at least no surprises :)

**Comment by NathanBWaters** [comment]
I really like @Aryk's suggestion, I'm doing the same thing currently with using comments to let others know backwards relationships.

**Comment by marxus85** [comment]
@NathanBWaters @Aryk as @plandem said, it's already an option. this is how i do it:

`
class Author extends Model {
  static modelName = 'Author'
  static fields = {
    id: attr(),
    title: attr(),
    books: many({
      to: 'Book',
      through: 'AuthorBook'
    })
  }
}

class Book extends Model {
  static modelName = 'Book'
  static fields = {
    id: attr(),
    title: attr(),
    authors: many({
      to: 'Author',
      through: 'AuthorBook'
    })
  }
}

class AuthorBook extends Model {
  static modelName = 'AuthorBook'
  static fields = {
    authorId: fk('Author'),
    bookId: fk('Book')
  }
}
`

no need to write alerting comments..

**Comment by Aryk** [comment]
@marxus85 - im refering to `fk` relationships. There is no way to define a one-to-many backwards on the other model currently...

**Comment by spachori94** [comment]
@plandem add ability to add indexes on columns (other than the ID, which is indexed by default) definitely needed

**Comment by michelalbers** [comment]
Is this project still alive? Love the simplicity and I think about throwing out realm js for it.

**Comment by nealoke** [comment]
@michelalbers I think it is, however even if it is not that active I already use it in various webapps and it is working like a charm :)

**Comment by markerikson** [comment]
Things have stalled a bit, but yeah, I think we ought to figure out what we really want for a 1.0 and get it out the door.  (#158 probably ought to be in there.)

**Comment by plandem** [comment]
library is using in production for few project at our company - quite huge projects with 20+ models/relations and 15k+ records for some models (not sure about 100k, but in some cases we have much more that 15k records).
